File: go/stream_session_test.go
package xrr_test

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"io"
	"path/filepath"
	"testing"

	xrr "hop.top/xrr"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func fixtureSession(t *testing.T, dir string) *xrr.FileSession {
	t.Helper()
	path := filepath.Join("..", "spec", "fixtures", dir)
	return xrr.NewSession(xrr.ModeReplay, xrr.NewFileCassette(path))
}

// grpcStreamOpen mirrors the gRPC adapter's open definition: canonical
// inputs service + method (+ msg_hash for content-addressed server
// streams), counter-addressed client/bidi, req payload {service, method}.
func grpcStreamOpen(typ xrr.StreamType, service, method string, msg []byte) xrr.StreamOpen {
	open := xrr.StreamOpen{
		AdapterID: "grpc",
		Type:      typ,
		Identity:  map[string]any{"service": service, "method": method},
		Payload:   map[string]any{"service": service, "method": method},
	}
	if typ == xrr.StreamServer {
		sum := sha256.Sum256(msg)
		open.Identity["msg_hash"] = hex.EncodeToString(sum[:4])
	} else {
		open.Counter = true
	}
	return open
}

// ── record path ────────────────────────────────────────────────────────────

func TestOpenStreamRecordServer(t *testing.T) {
	dir := t.TempDir()
	s := xrr.NewSession(xrr.ModeRecord, xrr.NewFileCassette(dir))
	msg := []byte(`{"path":"/etc/hosts"}`)
	rec, err := s.OpenStreamRecord(grpcStreamOpen(xrr.StreamServer, "files.FileService", "Download", msg))
	require.NoError(t, err)
	assert.Equal(t, "58a4bf3f", rec.Fingerprint())

	rec.RecordSend(msg)
	rec.RecordHalfClose()
	rec.RecordRecv([]byte("chunk-one\n"))
	rec.RecordRecv([]byte("chunk-two\n"))
	require.NoError(t, rec.Finish(map[string]any{"status_code": 0}, nil))

	pair, err := xrr.NewFileCassette(dir).LoadStream("grpc", "58a4bf3f")
	require.NoError(t, err)
	assert.Equal(t, xrr.StreamServer, pair.Req.Type)

	// Dense seq 0..N-1 counting all events in arrival order.
	require.Len(t, pair.Req.Frames, 1)
	assert.Equal(t, 0, pair.Req.Frames[0].Seq)
	require.NotNil(t, pair.Req.HalfClose)
	assert.Equal(t, 1, pair.Req.HalfClose.Seq)
	require.Len(t, pair.Resp.Frames, 2)
	assert.Equal(t, 2, pair.Resp.Frames[0].Seq)
	assert.Equal(t, 3, pair.Resp.Frames[1].Seq)
	assert.Equal(t, 4, pair.Resp.End.Seq)

	assert.Equal(t, []byte("chunk-one\n"), pair.Resp.Frames[0].Message)

	// at_ms stamped on every event, ≥ 0 and non-decreasing.
	prev := int64(0)
	for _, f := range append(append([]xrr.StreamFrame{}, pair.Req.Frames...), pair.Resp.Frames...) {
		require.NotNil(t, f.AtMs)
		assert.GreaterOrEqual(t, *f.AtMs, prev)
		prev = *f.AtMs
	}
	require.NotNil(t, pair.Req.HalfClose.AtMs)
	require.NotNil(t, pair.Resp.End.AtMs)

	// Server-stream payload carries no occurrence ordinal.
	assert.Equal(t, "files.FileService", pair.ReqPayload["service"])
	assert.Equal(t, "Download", pair.ReqPayload["method"])
	assert.NotContains(t, pair.ReqPayload, "n")
}

// TestOpenStreamRecordCounterN — one Session object is one counter domain:
// two opens of the same (service, method, type) tuple record n=0 then n=1,
// matching the grpc-client-stream-repeat fixture fingerprints.
func TestOpenStreamRecordCounterN(t *testing.T) {
	dir := t.TempDir()
	s := xrr.NewSession(xrr.ModeRecord, xrr.NewFileCassette(dir))
	open := grpcStreamOpen(xrr.StreamClient, "files.FileService", "Upload", nil)

	rec1, err := s.OpenStreamRecord(open)
	require.NoError(t, err)
	assert.Equal(t, "2bebfd6f", rec1.Fingerprint())
	rec1.RecordSend([]byte("alpha\n"))
	rec1.RecordHalfClose()
	rec1.RecordRecv([]byte(`{"received_bytes":6}`))
	require.NoError(t, rec1.Finish(map[string]any{"status_code": 0}, nil))

	rec2, err := s.OpenStreamRecord(open)
	require.NoError(t, err)
	assert.Equal(t, "b27b5fe1", rec2.Fingerprint())
	rec2.RecordHalfClose()
	require.NoError(t, rec2.Finish(map[string]any{"status_code": 0}, nil))

	c := xrr.NewFileCassette(dir)
	p1, err := c.LoadStream("grpc", "2bebfd6f")
	require.NoError(t, err)
	assert.Equal(t, 0, p1.ReqPayload["n"])
	p2, err := c.LoadStream("grpc", "b27b5fe1")
	require.NoError(t, err)
	assert.Equal(t, 1, p2.ReqPayload["n"])

	// A different tuple starts its own count.
	rec3, err := s.OpenStreamRecord(grpcStreamOpen(xrr.StreamBidi, "chat.ChatService", "Converse", nil))
	require.NoError(t, err)
	assert.Equal(t, "c6233d2e", rec3.Fingerprint())
}

// TestStreamRecordingTerminalIsFinal — no events are recorded after the
// terminal; a second Finish is an error.
func TestStreamRecordingTerminalIsFinal(t *testing.T) {
	dir := t.TempDir()
	s := xrr.NewSession(xrr.ModeRecord, xrr.NewFileCassette(dir))
	rec, err := s.OpenStreamRecord(grpcStreamOpen(xrr.StreamBidi, "chat.ChatService", "Converse", nil))
	require.NoError(t, err)
	rec.RecordSend([]byte("ping-1"))
	rec.RecordRecv([]byte("pong-1"))
	require.NoError(t, rec.Finish(map[string]any{"status_code": 0}, nil))

	// Dropped, matching the real-world no-op.
	rec.RecordSend([]byte("late"))
	rec.RecordRecv([]byte("late"))
	rec.RecordHalfClose()
	assert.Error(t, rec.Finish(map[string]any{"status_code": 0}, nil))

	pair, err := xrr.NewFileCassette(dir).LoadStream("grpc", rec.Fingerprint())
	require.NoError(t, err)
	assert.Len(t, pair.Req.Frames, 1)
	assert.Len(t, pair.Resp.Frames, 1)
	assert.Nil(t, pair.Req.HalfClose)
	assert.Equal(t, 2, pair.Resp.End.Seq)
}

func TestStreamRecordingErrorTerminal(t *testing.T) {
	dir := t.TempDir()
	s := xrr.NewSession(xrr.ModeRecord, xrr.NewFileCassette(dir))
	rec, err := s.OpenStreamRecord(grpcStreamOpen(
		xrr.StreamServer, "files.FileService", "Download", []byte(`{"path":"/var/log/big.log"}`)))
	require.NoError(t, err)
	rec.RecordSend([]byte(`{"path":"/var/log/big.log"}`))
	rec.RecordHalfClose()
	rec.RecordRecv([]byte("log-chunk-1\n"))
	require.NoError(t, rec.Finish(
		map[string]any{"status_code": 14},
		errors.New("rpc error: code = Unavailable desc = connection reset"),
	))

	pair, err := xrr.NewFileCassette(dir).LoadStream("grpc", "9e8c4d4c")
	require.NoError(t, err)
	assert.Equal(t, "rpc error: code = Unavailable desc = connection reset", pair.RecordedErr)
	assert.Equal(t, 14, pair.RespPayload["status_code"])
}

// ── replay path ────────────────────────────────────────────────────────────

func TestOpenStreamReplayBidi(t *testing.T) {
	s := fixtureSession(t, "grpc-bidi-stream")
	rep, err := s.OpenStreamReplay(grpcStreamOpen(xrr.StreamBidi, "chat.ChatService", "Converse", nil))
	require.NoError(t, err)
	assert.Equal(t, "c6233d2e", rep.Fingerprint())
	assert.Equal(t, xrr.StreamBidi, rep.Type())

	// Reads never gate on send progress: drain both pongs first.
	msg, err := rep.Recv()
	require.NoError(t, err)
	assert.Equal(t, []byte("pong-1"), msg)
	msg, err = rep.Recv()
	require.NoError(t, err)
	assert.Equal(t, []byte("pong-2"), msg)

	// Sends validated in order and bytes afterwards.
	require.NoError(t, rep.Send([]byte("ping-1")))
	require.NoError(t, rep.Send([]byte("ping-2")))
	require.NoError(t, rep.HalfClose())

	_, err = rep.Recv()
	assert.ErrorIs(t, err, io.EOF)
	_, err = rep.Recv()
	assert.ErrorIs(t, err, io.EOF, "terminal repeats for j > R")
}

func TestStreamReplaySendMismatchIsTerminal(t *testing.T) {
	s := fixtureSession(t, "grpc-bidi-stream")
	rep, err := s.OpenStreamReplay(grpcStreamOpen(xrr.StreamBidi, "chat.ChatService", "Converse", nil))
	require.NoError(t, err)

	require.NoError(t, rep.Send([]byte("ping-1")))
	err = rep.Send([]byte("ping-DIVERGED"))
	require.Error(t, err)
	assert.ErrorIs(t, err, xrr.ErrStreamMismatch)
	var mErr *xrr.StreamMismatchError
	require.ErrorAs(t, err, &mErr)
	assert.Equal(t, 1, mErr.Ordinal)

	// Mismatch poisons every subsequent operation.
	_, err = rep.Recv()
	assert.ErrorIs(t, err, xrr.ErrStreamMismatch)
	assert.ErrorIs(t, rep.HalfClose(), xrr.ErrStreamMismatch)
	assert.ErrorIs(t, rep.Send([]byte("ping-2")), xrr.ErrStreamMismatch)
}

func TestStreamReplayShortHalfCloseIsMismatch(t *testing.T) {
	s := fixtureSession(t, "grpc-client-stream")
	rep, err := s.OpenStreamReplay(grpcStreamOpen(xrr.StreamClient, "files.FileService", "Upload", nil))
	require.NoError(t, err)

	require.NoError(t, rep.Send([]byte("part-one\n")))
	err = rep.HalfClose()
	assert.ErrorIs(t, err, xrr.ErrStreamMismatch, "half-close after 1 of 2 sends")
}

// TestStreamReplayPostCompletionSend — send at i ≥ S with an OK terminal is
// the non-poisoning stream-done signal; the recv side is unaffected.
func TestStreamReplayPostCompletionSend(t *testing.T) {
	s := fixtureSession(t, "grpc-client-stream")
	rep, err := s.OpenStreamReplay(grpcStreamOpen(xrr.StreamClient, "files.FileService", "Upload", nil))
	require.NoError(t, err)

	require.NoError(t, rep.Send([]byte("part-one\n")))
	require.NoError(t, rep.Send([]byte("part-two\n")))
	assert.ErrorIs(t, rep.Send([]byte("part-three\n")), io.EOF)
	require.NoError(t, rep.HalfClose(), "half-close after all recorded sends is always accepted")

	msg, err := rep.Recv()
	require.NoError(t, err)
	assert.Equal(t, []byte(`{"received_bytes":18}`), msg)
	_, err = rep.Recv()
	assert.ErrorIs(t, err, io.EOF)
}

// TestStreamReplayMidStreamError — recorded frames delivered, then the
// recorded error in place of end-of-stream; post-completion sends surface
// the same recorded error.
func TestStreamReplayMidStreamError(t *testing.T) {
	s := fixtureSession(t, "grpc-stream-error")
	rep, err := s.OpenStreamReplay(grpcStreamOpen(
		xrr.StreamServer, "files.FileService", "Download", []byte(`{"path":"/var/log/big.log"}`)))
	require.NoError(t, err)
	assert.Equal(t, "9e8c4d4c", rep.Fingerprint())
	assert.Equal(t, 14, rep.RespPayload()["status_code"])

	require.NoError(t, rep.Send([]byte(`{"path":"/var/log/big.log"}`)))
	require.NoError(t, rep.HalfClose())

	msg, err := rep.Recv()
	require.NoError(t, err)
	assert.Equal(t, []byte("log-chunk-1\n"), msg)
	msg, err = rep.Recv()
	require.NoError(t, err)
	assert.Equal(t, []byte("log-chunk-2\n"), msg)

	wantErr := "rpc error: code = Unavailable desc = connection reset"
	_, err = rep.Recv()
	require.EqualError(t, err, wantErr)
	assert.NotErrorIs(t, err, xrr.ErrStreamMismatch)
	_, err = rep.Recv()
	require.EqualError(t, err, wantErr, "recorded error repeats for j > R")

	// The recorded stream was already dead: post-completion send returns it.
	assert.EqualError(t, rep.Send([]byte("extra")), wantErr)
}

func TestStreamReplayEmptyStreams(t *testing.T) {
	t.Run("server empty resp", func(t *testing.T) {
		s := fixtureSession(t, "grpc-stream-empty")
		rep, err := s.OpenStreamReplay(grpcStreamOpen(
			xrr.StreamServer, "files.FileService", "Download", []byte(`{"path":"/etc/empty"}`)))
		require.NoError(t, err)
		_, err = rep.Recv()
		assert.ErrorIs(t, err, io.EOF, "first read yields end-of-stream")
	})

	t.Run("client immediate half-close", func(t *testing.T) {
		s := fixtureSession(t, "grpc-stream-empty")
		rep, err := s.OpenStreamReplay(grpcStreamOpen(xrr.StreamClient, "telemetry.MetricsService", "Push", nil))
		require.NoError(t, err)
		require.NoError(t, rep.HalfClose(), "S=0: immediate half-close accepted")
		msg, err := rep.Recv()
		require.NoError(t, err)
		assert.Equal(t, []byte(`{"count":0}`), msg)
		_, err = rep.Recv()
		assert.ErrorIs(t, err, io.EOF)
	})

	t.Run("bidi no traffic", func(t *testing.T) {
		s := fixtureSession(t, "grpc-stream-empty")
		rep, err := s.OpenStreamReplay(grpcStreamOpen(xrr.StreamBidi, "chat.ChatService", "Ping", nil))
		require.NoError(t, err)
		require.NoError(t, rep.HalfClose())
		_, err = rep.Recv()
		assert.ErrorIs(t, err, io.EOF)
	})
}

func TestStreamReplayMissAndShapeMismatch(t *testing.T) {
	dir := t.TempDir()
	c := xrr.NewFileCassette(dir)
	s := xrr.NewSession(xrr.ModeReplay, c)

	// No pair on disk ⇒ cassette miss.
	_, err := s.OpenStreamReplay(grpcStreamOpen(xrr.StreamBidi, "s", "m", nil))
	assert.ErrorIs(t, err, xrr.ErrCassetteMiss)

	// A unary pair at the streamed fingerprint ⇒ shape mismatch, not a miss.
	fp, err := xrr.StreamFingerprint(grpcStreamOpen(xrr.StreamBidi, "s", "m", nil), 1)
	require.NoError(t, err)
	require.NoError(t, c.Save("grpc", fp,
		map[string]any{"service": "s", "method": "m"},
		map[string]any{"status_code": 0}, nil))
	_, err = s.OpenStreamReplay(grpcStreamOpen(xrr.StreamBidi, "s", "m", nil))
	assert.ErrorIs(t, err, xrr.ErrShapeMismatch)
	assert.NotErrorIs(t, err, xrr.ErrCassetteMiss)
}

func TestOpenStreamModeEnforcement(t *testing.T) {
	dir := t.TempDir()
	open := grpcStreamOpen(xrr.StreamBidi, "s", "m", nil)

	replaySession := xrr.NewSession(xrr.ModeReplay, xrr.NewFileCassette(dir))
	assert.Equal(t, xrr.ModeReplay, replaySession.Mode())
	_, err := replaySession.OpenStreamRecord(open)
	assert.Error(t, err)

	recordSession := xrr.NewSession(xrr.ModeRecord, xrr.NewFileCassette(dir))
	_, err = recordSession.OpenStreamReplay(open)
	assert.Error(t, err)
}

File: ts/tests/streamSession.test.ts
import fs from "node:fs";
import path from "node:path";
import { describe, expect, test } from "vitest";
import { FileCassette } from "../src/cassette.js";
import { FileSession } from "../src/session.js";
import { type StreamType, type StreamedInteraction } from "../src/stream.js";
import { streamFingerprint } from "../src/streamfp.js";
import { ErrEndOfStream } from "../src/streamSession.js";
import { type Cassette } from "../src/xrr.js";
import {
  caught,
  fixturesRoot,
  grpcStreamOpen,
  sseStreamOpen,
  td,
  tmpDir,
  utf8,
} from "./streamSessionHelpers.js";

// ── identity seam ──────────────────────────────────────────────────────────

describe("streamFingerprint — identity seam", () => {
  test("reproduces the gRPC spec vectors", () => {
    expect(
      streamFingerprint(
        grpcStreamOpen("server", "files.FileService", "Download", utf8('{"path":"/etc/hosts"}')),
        -1
      )
    ).toBe("58a4bf3f");
    expect(
      streamFingerprint(
        grpcStreamOpen(
          "server",
          "files.FileService",
          "Download",
          utf8('{"path":"/var/log/big.log"}')
        ),
        -1
      )
    ).toBe("9e8c4d4c");
    expect(
      streamFingerprint(grpcStreamOpen("client", "files.FileService", "Upload"), 0)
    ).toBe("2bebfd6f");
    expect(
      streamFingerprint(grpcStreamOpen("client", "files.FileService", "Upload"), 1)
    ).toBe("b27b5fe1");
    expect(
      streamFingerprint(grpcStreamOpen("bidi", "chat.ChatService", "Converse"), 0)
    ).toBe("c6233d2e");
  });

  test("url-keyed identity reproduces the sse fixture fingerprint", () => {
    expect(streamFingerprint(sseStreamOpen("https://example.test/events"), -1)).toBe("66ecc77a");
  });

  test("rejects reserved identity keys", () => {
    const base = sseStreamOpen("https://example.test/events");
    expect(() =>
      streamFingerprint({ ...base, identity: { url: "u", stream: "server" } }, -1)
    ).toThrow(/reserved/);
    expect(() => streamFingerprint({ ...base, identity: { url: "u", n: 3 } }, -1)).toThrow(
      /reserved/
    );
  });

  test("rejects an invalid stream type", () => {
    const open = { ...sseStreamOpen("u"), type: "duplex" as StreamType };
    expect(() => streamFingerprint(open, -1)).toThrow(/invalid/);
  });

  test("counter-addressed opens require n >= 0", () => {
    expect(() =>
      streamFingerprint(grpcStreamOpen("client", "files.FileService", "Upload"), -1)
    ).toThrow(/n must be >= 0/);
  });
});

// ── record path ────────────────────────────────────────────────────────────

describe("openStreamRecord", () => {
  test("server round-trip: pair on disk validates, dense seq, at_ms monotonic", async () => {
    const dir = tmpDir();
    const session = new FileSession("record", new FileCassette(dir));
    const msg = utf8('{"path":"/etc/hosts"}');
    const rec = await session.openStreamRecord(
      grpcStreamOpen("server", "files.FileService", "Download", msg)
    );
    expect(rec.fingerprint).toBe("58a4bf3f");

    rec.recordSend(msg);
    rec.recordHalfClose();
    rec.recordRecv(utf8("chunk-one\n"));
    rec.recordRecv(utf8("chunk-two\n"));
    await rec.finish({ status_code: 0 });

    // Fingerprint filenames per the v1 naming convention.
    expect(fs.existsSync(path.join(dir, "grpc-58a4bf3f.req.yaml"))).toBe(true);
    expect(fs.existsSync(path.join(dir, "grpc-58a4bf3f.resp.yaml"))).toBe(true);

    // loadStreamed validates the pair (ascending frames, seq uniqueness,
    // maximal end.seq) — a load without throwing is a validity check itself.
    const pair = await new FileCassette(dir).loadStreamed("grpc", "58a4bf3f");
    expect(pair.req.stream.type).toBe("server");

    // Dense seq 0..N-1 counting all events in arrival order.
    expect(pair.req.stream.frames.map((f) => f.seq)).toEqual([0]);
    expect(pair.req.stream.half_close?.seq).toBe(1);
    expect(pair.resp.stream.frames.map((f) => f.seq)).toEqual([2, 3]);
    expect(pair.resp.stream.end.seq).toBe(4);

    expect(td.decode(pair.resp.stream.frames[0].message)).toBe("chunk-one\n");
    expect(td.decode(pair.resp.stream.frames[1].message)).toBe("chunk-two\n");

    // at_ms stamped on every event, ≥ 0 and non-decreasing in event order.
    const atMs = [
      pair.req.stream.frames[0].at_ms,
      pair.req.stream.half_close?.at_ms,
      pair.resp.stream.frames[0].at_ms,
      pair.resp.stream.frames[1].at_ms,
      pair.resp.stream.end.at_ms,
    ];
    for (const at of atMs) expect(at).toBeGreaterThanOrEqual(0);
    for (let i = 1; i < atMs.length; i++) {
      expect(atMs[i]).toBeGreaterThanOrEqual(atMs[i - 1] as number);
    }

    // Server-stream payload carries no occurrence ordinal.
    expect(pair.req.payload).toEqual({ service: "files.FileService", method: "Download" });
  });

  test("recorded bidi conversation matches the fixture shape", async () => {
    const dir = tmpDir();
    const session = new FileSession("record", new FileCassette(dir));
    const rec = await session.openStreamRecord(
      grpcStreamOpen("bidi", "chat.ChatService", "Converse")
    );
    rec.recordSend(utf8("ping-1"));
    rec.recordRecv(utf8("pong-1"));
    rec.recordSend(utf8("ping-2"));
    rec.recordRecv(utf8("pong-2"));
    rec.recordHalfClose();
    await rec.finish({ status_code: 0 });

    const shape = (pair: StreamedInteraction) => ({
      type: pair.req.stream.type,
      reqPayload: pair.req.payload,
      respPayload: pair.resp.payload,
      error: pair.resp.error ?? "",
      sends: pair.req.stream.frames.map((f) => ({ seq: f.seq, text: td.decode(f.message) })),
      recvs: pair.resp.stream.frames.map((f) => ({ seq: f.seq, text: td.decode(f.message) })),
      halfClose: pair.req.stream.half_close?.seq ?? null,
      end: pair.resp.stream.end.seq,
    });

    const recorded = await new FileCassette(dir).loadStreamed("grpc", "c6233d2e");
    const fixture = await new FileCassette(
      path.join(fixturesRoot, "grpc-bidi-stream")
    ).loadStreamed("grpc", "c6233d2e");
    expect(shape(recorded)).toEqual(shape(fixture));
  });

  test("counter-addressed opens record n=0 then n=1 in one session", async () => {
    const dir = tmpDir();
    const session = new FileSession("record", new FileCassette(dir));
    const open = () => grpcStreamOpen("client", "files.FileService", "Upload");

    const rec1 = await session.openStreamRecord(open());
    expect(rec1.fingerprint).toBe("2bebfd6f");
    rec1.recordSend(utf8("alpha\n"));
    rec1.recordHalfClose();
    rec1.recordRecv(utf8('{"received_bytes":6}'));
    await rec1.finish({ status_code: 0 });

    const rec2 = await session.openStreamRecord(open());
    expect(rec2.fingerprint).toBe("b27b5fe1");
    rec2.recordHalfClose();
    await rec2.finish({ status_code: 0 });

    // The informational occurrence ordinal is recoverable from disk.
    const cassette = new FileCassette(dir);
    const p1 = await cassette.loadStreamed("grpc", "2bebfd6f");
    expect((p1.req.payload as { n?: number }).n).toBe(0);
    const p2 = await cassette.loadStreamed("grpc", "b27b5fe1");
    expect((p2.req.payload as { n?: number }).n).toBe(1);

    // A different tuple starts its own count.
    const rec3 = await session.openStreamRecord(
      grpcStreamOpen("bidi", "chat.ChatService", "Converse")
    );
    expect(rec3.fingerprint).toBe("c6233d2e");
  });

  test("post-finish events are dropped; a second finish rejects", async () => {
    const dir = tmpDir();
    const session = new FileSession("record", new FileCassette(dir));
    const rec = await session.openStreamRecord(
      grpcStreamOpen("bidi", "chat.ChatService", "Converse")
    );
    rec.recordSend(utf8("ping-1"));
    rec.recordRecv(utf8("pong-1"));
    await rec.finish({ status_code: 0 });

    // Dropped, matching the real-world no-op.
    rec.recordSend(utf8("late"));
    rec.recordRecv(utf8("late"));
    rec.recordHalfClose();
    await expect(rec.finish({ status_code: 0 })).rejects.toThrow(/already finished/);

    const pair = await new FileCassette(dir).loadStreamed("grpc", rec.fingerprint);
    expect(pair.req.stream.frames).toHaveLength(1);
    expect(pair.resp.stream.frames).toHaveLength(1);
    expect(pair.req.stream.half_close).toBeUndefined();
    expect(pair.resp.stream.end.seq).toBe(2);
  });

  test("error terminal persists the envelope error and payload", async () => {
    const dir = tmpDir();
    const session = new FileSession("record", new FileCassette(dir));
    const msg = utf8('{"path":"/var/log/big.log"}');
    const rec = await session.openStreamRecord(
      grpcStreamOpen("server", "files.FileService", "Download", msg)
    );
    rec.recordSend(msg);
    rec.recordHalfClose();
    rec.recordRecv(utf8("log-chunk-1\n"));
    await rec.finish(
      { status_code: 14 },
      new Error("rpc error: code = Unavailable desc = connection reset")
    );

    const pair = await new FileCassette(dir).loadStreamed("grpc", "9e8c4d4c");
    expect(pair.resp.error).toBe("rpc error: code = Unavailable desc = connection reset");
    expect(pair.resp.payload).toEqual({ status_code: 14 });
  });

  test("empty recording emits frames: [] on both sides and validates", async () => {
    const dir = tmpDir();
    const session = new FileSession("record", new FileCassette(dir));
    const rec = await session.openStreamRecord(
      grpcStreamOpen("bidi", "chat.ChatService", "Ping")
    );
    rec.recordHalfClose();
    await rec.finish({ status_code: 0 });

    const pair = await new FileCassette(dir).loadStreamed("grpc", "ebbd3938");
    expect(pair.req.stream.frames).toEqual([]);
    expect(pair.resp.stream.frames).toEqual([]);
    expect(pair.req.stream.half_close?.seq).toBe(0);
    expect(pair.resp.stream.end.seq).toBe(1);
  });
});

// ── open guards ────────────────────────────────────────────────────────────

describe("stream open guards", () => {
  test("mode enforcement in both directions", async () => {
    const dir = tmpDir();
    const open = grpcStreamOpen("bidi", "s", "m");
    const replaySession = new FileSession("replay", new FileCassette(dir));
    await expect(replaySession.openStreamRecord(open)).rejects.toThrow(/requires record mode/);
    const recordSession = new FileSession("record", new FileCassette(dir));
    await expect(recordSession.openStreamReplay(open)).rejects.toThrow(/requires replay mode/);
  });

  test("adapter id is required", async () => {
    const session = new FileSession("record", new FileCassette(tmpDir()));
    const open = { ...grpcStreamOpen("bidi", "s", "m"), adapterID: "" };
    await expect(session.openStreamRecord(open)).rejects.toThrow(/adapter id/);
  });

  test("a cassette without streamed IO is rejected", async () => {
    const unary: Cassette = {
      save: async () => {},
      load: async () => ({ req: {}, resp: {} }),
    };
    const session = new FileSession("record", unary);
    await expect(
      session.openStreamRecord(grpcStreamOpen("bidi", "s", "m"))
    ).rejects.toThrow(/stream-capable/);
  });
});

// ── record → replay integration ────────────────────────────────────────────

describe("record then replay", () => {
  test("a recorded pair replays through a fresh replay session", async () => {
    const dir = tmpDir();
    const recSession = new FileSession("record", new FileCassette(dir));
    const rec = await recSession.openStreamRecord(
      grpcStreamOpen("client", "files.FileService", "Upload")
    );
    rec.recordSend(utf8("part-one\n"));
    rec.recordSend(utf8("part-two\n"));
    rec.recordHalfClose();
    rec.recordRecv(utf8('{"received_bytes":18}'));
    await rec.finish({ status_code: 0 });

    const repSession = new FileSession("replay", new FileCassette(dir));
    const rep = await repSession.openStreamReplay(
      grpcStreamOpen("client", "files.FileService", "Upload")
    );
    expect(rep.fingerprint).toBe("2bebfd6f");
    rep.send(utf8("part-one\n"));
    rep.send(utf8("part-two\n"));
    rep.halfClose();
    expect(td.decode(rep.recv())).toBe('{"received_bytes":18}');
    expect(caught(() => rep.recv())).toBe(ErrEndOfStream);
  });
});
